sql/wsrep_mysqld.cc

```diff
@@ -1665,86 +1666,82 @@ wsrep::key_array wsrep_prepare_keys_for_toi(const char *db, const char *table,
 /*
  * Collect wsrep keys corresponding to tables that are referencing given table
  */
-static bool collect_fk_children(const dd::Table *table_def,
+static void collect_fk_children(const dd::Table *table_def,
                                 wsrep::key_array *keys) {
-  for (const dd::Foreign_key_parent *fk : table_def->foreign_key_parents()) {
+  for (auto fk : table_def->foreign_key_parents()) {
     keys->push_back(wsrep_prepare_key_for_toi(fk->child_schema_name().c_str(),
                                               fk->child_table_name().c_str(),
                                               wsrep::key::shared));
   }
-  return false;
 }
 
 void wsrep_append_child_tables(THD *thd, TABLE_LIST *tables,
                                wsrep::key_array *keys) {
   if (!WSREP(thd) || !WSREP_CLIENT(thd)) return;
-  TABLE_LIST *table;
-
-  thd->mdl_context.release_transactional_locks();
-  MDL_savepoint mdl_savepoint = thd->mdl_context.mdl_savepoint();
 
-  uint counter;
-  if (open_temporary_tables(thd, tables) ||
-      open_tables(thd, &tables, &counter,
-                  MYSQL_OPEN_FORCE_SHARED_HIGH_PRIO_MDL)) {
-    WSREP_DEBUG("unable to open table for FK checks for %s", thd->query().str);
-  }
+  for (auto table = tables; table; table = table->next_local) {
+    if (!is_temporary_table(table)) {
+      MDL_ticket *mdl_ticket = nullptr;
+      if (dd::acquire_shared_table_mdl(thd, table->db, table->table_name, false,
+                                       &mdl_ticket)) {
+        assert(0);
+        continue;
+      }
 
-  for (table = tables; table; table = table->next_local) {
-    if (!is_temporary_table(table) && table->table) {
       const dd::Table *table_obj = nullptr;
       dd::cache::Dictionary_client::Auto_releaser releaser(thd->dd_client());
-      if (thd->dd_client()->acquire(
-              dd::String_type(table->table->s->db.str),
-              dd::String_type(table->table->s->table_name.str), &table_obj)) {
+      if (thd->dd_client()->acquire(dd::String_type(table->db),
+                                    dd::String_type(table->table_name),
+                                    &table_obj)) {
         assert(0);
+        dd::release_mdl(thd, mdl_ticket);
+        continue;
       }
       collect_fk_children(table_obj, keys);
```

Comment:
Can we add an assert that `table_obj` ptr is valid here?
```suggestion
      }
      assert(tab_obj);
      collect_fk_children(table_obj, keys);
```

Reply:
No, because it can be nullptr if the table does not exist. That's why we check if it is not nullptr (inside collect_fk* before refactoring, and right before calling action() after refactoring.